Implement process related tables on FreeBSD.

wxsBSD opened the pull request:
This implements the following tables on FreeBSD:

process_envs
process_memory_map
process_open_files
process_open_sockets
processes

All the heavy lifting is done with libprocstat(3). All the tables follow
the same general principle. Use the common function, getProcesses() in
procstat.cpp, to get the processes and then generate the rows for each
process returned. There is also a procstatCleanup() function commonly
used across all the tables.

The one thing I am not able to test is the process_open_sockets table on
an IPv6 machine.

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

wxsBSD commented:
If anyone has an IPv6 machine they can test out the ```process_open_sockets``` table on I would owe you a beverage of your choice. I'm trying to get my hands on one but it may be easier to see if @jacknagz has one. :)

theopolis commented:
I hope someone is tallying the drink owing dependency list. August is coming...

Before I have a chance to look into the code-- is there an AWS AMI we test with IPv6 or perhaps some additional Vagrant commands to set up VMs with IPv6 local nets? Maybe a network between host and guest would provide some observable socket opens?

wxsBSD commented:
Now I'm curious what is happening in August? I suspect I owe more than I realize! ;)

I was doing my testing by connecting to local host over IPv4, I see no reason why I can't do the same over IPv6. I'll add an IPv6 address to an interface and try it out. I'll let you know how to do that if it works.

zi0r commented:
:)

theopolis commented on osquery/tables/networking/freebsd/process_open_sockets.cpp:
```diff
@@ -0,0 +1,121 @@
+/*
+ *  Copyright (c) 2015, Wesley Shields.
```
These copyright notices will be overwritten by bots and codemodes to 2014, Facebook as part of the CLA. If we want names/handles to persist past git history we should use a CONTRIBUTORS file.

theopolis commented on osquery/tables/networking/freebsd/process_open_sockets.cpp:
```diff
@@ -0,0 +1,121 @@
+// Heavily inspired by procstat(1) on FreeBSD.
+std::pair<std::string, int>& sockaddr_to_pair(struct sockaddr_storage* sstor) {
+  char buffer[INET6_ADDRSTRLEN] = {0};
+  struct sockaddr_in6* sin6;
```
Nit, mind initializing these to `nullptr`, it helps a lot with static analysis later on.

theopolis commented on osquery/tables/networking/freebsd/process_open_sockets.cpp:
```diff
@@ -0,0 +1,121 @@
+  std::pair<std::string, int> addr;
+
+  files = procstat_getfiles(pstat, proc, 0);
+  if (files != nullptr) {
```
Depending on your philosophy of program flow. You could return in the opposite condition and the following block would be more left-heavy.

wxsBSD replied:
I generally prefer to do what you are talking about, to avoid excessive indentation. I'll fix this along with the other things you mention over lunch. :)

theopolis commented on osquery/tables/system/freebsd/procstat.h:
```diff
@@ -0,0 +1,32 @@
+ *
+ */
+
+#ifndef OSQUERY_TABLES_SYSTEM_PROCSTAT_H
```
You can use `#pragma once` instead of the ifndef/define logic.

wxsBSD commented:
I added an IPv6 address to the interface in my jail and made sure sshd was listening on it. Works like a champ. :)

```
osquery% sudo procstat -f 54790                                                                             
  PID COMM                FD T V FLAGS    REF  OFFSET PRO NAME        
54790 sshd              text v r r-------   -       - -   /usr/sbin/sshd    
54790 sshd               cwd v d r-------   -       - -   /                 
54790 sshd              root v d r-------   -       - -   /                 
54790 sshd              jail v d r-------   -       - -   /                 
54790 sshd                 0 v c rw------   3       0 -   /dev/null         
54790 sshd                 1 v c rw------   3       0 -   /dev/null         
54790 sshd                 2 v c rw------   3       0 -   /dev/null         
54790 sshd                 3 s - rw---n--   1       0 TCP fd3e:9e6:ccf8:1291::2.22 ::.0
osquery% sudo ./build/debug_11.0/osquery/osqueryi "select * from process_open_sockets where pid=54790;"
+-------+--------+--------+----------+-----------------------+----------------+------------+-------------+
| pid   | socket | family | protocol | local_address         | remote_address | local_port | remote_port |
+-------+--------+--------+----------+-----------------------+----------------+------------+-------------+
| 54790 | 3      | 28     | 6        | fd3e:9e6:ccf8:1291::2 | ::             | 22         | 0           |
+-------+--------+--------+----------+-----------------------+----------------+------------+-------------+
osquery%
```

The same thing should work for you. Just make sure your interface has an IPv6 address and that sshd is listening on it and you should be fine.

theopolis commented on osquery/tables/CMakeLists.txt:
```diff
@@ -24,6 +24,11 @@ elseif(FREEBSD)
   ADD_OSQUERY_TABLE_TEST(${OSQUERY_FREEBSD_TABLES_TESTS})
+
+  ADD_OSQUERY_LINK_ADDITIONAL("procstat")
+  ADD_OSQUERY_LINK_ADDITIONAL("util")
```
Are `libutil`, `libkvm`, `libelf` required by a static procstat?

wxsBSD replied:
Yes, they are.

theopolis commented on osquery/tables/networking/freebsd/process_open_sockets.cpp:
```diff
@@ -0,0 +1,122 @@
+      continue;
+    }
+
+    error = procstat_get_socket_info(pstat, file, &sock, NULL);
```
nit, `nullptr` for the `char *errbuf` 4th param.

theopolis commented on osquery/tables/networking/freebsd/process_open_sockets.cpp:
```diff
@@ -0,0 +1,122 @@
+  return addr;
+}
+
+void genSockets(struct procstat* pstat,
```
Can `pstat` and `proc` be pass const?

wxsBSD replied:
Nope. ```procstat_getfiles``` and friends don't expect them to be const and clang complains.

theopolis commented on osquery/tables/networking/freebsd/process_open_sockets.cpp:
```diff
@@ -0,0 +1,122 @@
+  procstat_freefiles(pstat, files);
+}
+
+QueryData genOpenSockets(QueryContext &context) {
```
If you wanted to get some C++11 points:
```cpp
QueryData results;
struct kinfo_proc* procs = nullptr;
struct procstat* pstat = nullptr;
auto cnt = getProcesses(context, &pstat, &procs);
for (size_t i = 0; i < cnt; i++) {
  genSockets(pstat, &procs[i], results);
}
[...]
```

theopolis commented on osquery/tables/system/freebsd/process_open_files.cpp:
```diff
@@ -0,0 +1,71 @@
+namespace osquery {
+namespace tables {
+
+void genDescriptors(struct procstat* pstat,
```
I'm not sure host the procstat APIs work, but in most cases can the `struct procstat*` params be solely input and passed as `const` in these gen functions?

theopolis commented on osquery/tables/system/freebsd/process_open_files.cpp:
```diff
@@ -0,0 +1,71 @@
+                    QueryData& results) {
+
+  Row r;
+  struct filestat_list* files;
```
More nits, `nullptr`

theopolis commented on osquery/tables/system/freebsd/processes.cpp:
```diff
@@ -8,33 +8,238 @@
+      if (vmentry[i].kve_vn_fileid == 0 && r["path"].size() > 0) {
+        r["pseudo"] = INTEGER("1");
+      }
+      else {
```
nit, no `\n`

theopolis commented on osquery/tables/system/freebsd/processes.cpp:
```diff
@@ -8,33 +8,238 @@
+      r["end"] = "0x" + TEXT(addr_str);
+
+      std::string permissions;
+      permissions += vmentry[i].kve_protection & KVME_PROT_READ ? "r" : "-";
```
This is great! Only nit is to wrap the operation in `([...])`.

theopolis commented on osquery/tables/system/freebsd/processes.cpp:
```diff
@@ -8,33 +8,238 @@
+  }
+}
+
+void genProcess(struct procstat* pstat,
```
Note above about `const`

theopolis commented on osquery/tables/system/freebsd/processes.cpp:
```diff
@@ -8,33 +8,238 @@
+  Row r;
+  static char path[PATH_MAX];
+  char** args;
+  struct filestat_list* files;
```
nit, `nullptr` initialization

theopolis commented:
Looks great @wxsBSD!

ok to test (although nothing will happen since we don't have a FreeBSD builder).

theopolis commented:
ok to test

wxsBSD commented:
I've addressed all the concerns mentioned except the ```const``` one which I don't think can be addressed. Let me know if you want me to squash this and I'll figure out how to do it. ;)

theopolis commented:
Yes please! Squash and we'll merge :)

wxsBSD commented:
![](http://www.foodsubs.com/Photos/squash-yellowcrookneck.jpg)'ed
